lib/galaxy/datatypes/images.py

```diff
@@ -104,17 +215,95 @@ class Tiff(Image):
     def set_meta(
         self, dataset: DatasetProtocol, overwrite: bool = True, metadata_tmp_files_dir: Optional[str] = None, **kwd
     ) -> None:
+        """
+        Populate the metadata of the TIFF image using the tifffile library.
+        """
         spec_key = "offsets"
-        offsets_file = dataset.metadata.offsets
-        if not offsets_file:
-            offsets_file = dataset.metadata.spec[spec_key].param.new_file(
-                dataset=dataset, metadata_tmp_files_dir=metadata_tmp_files_dir
-            )
-        with tifffile.TiffFile(dataset.get_file_name()) as tif:
-            offsets = [page.offset for page in tif.pages]
-        with open(offsets_file.get_file_name(), "w") as f:
-            json.dump(offsets, f)
-        dataset.metadata.offsets = offsets_file
+        if hasattr(dataset.metadata, spec_key):
+            offsets_file = dataset.metadata.offsets
+            if not offsets_file:
+                offsets_file = dataset.metadata.spec[spec_key].param.new_file(
+                    dataset=dataset, metadata_tmp_files_dir=metadata_tmp_files_dir
+                )
+        else:
+            offsets_file = None
+        try:
+            with tifffile.TiffFile(dataset.get_file_name()) as tif:
+                offsets = [page.offset for page in tif.pages]
+
+                # Aggregate a list of values for each metadata field (one value for each page of the TIFF file)
+                metadata: Dict[str, List[Any]] = {
+                    key: []
+                    for key in [
+                        "axes",
+                        "dtype",
+                        "width",
+                        "height",
+                        "channels",
+                        "depth",
+                        "frames",
+                        "num_unique_values",
+                    ]
+                }
+                for page in tif.series:
+
+                    # Determine the metadata values that should be generally available
+                    metadata["axes"].append(page.axes.upper())
+                    metadata["dtype"].append(str(page.dtype))
+
+                    axes = metadata["axes"][-1].replace("S", "C")
+                    metadata["width"].append(Tiff._get_axis_size(page.shape, axes, "X"))
+                    metadata["height"].append(Tiff._get_axis_size(page.shape, axes, "Y"))
+                    metadata["channels"].append(Tiff._get_axis_size(page.shape, axes, "C"))
+                    metadata["depth"].append(Tiff._get_axis_size(page.shape, axes, "Z"))
+                    metadata["frames"].append(Tiff._get_axis_size(page.shape, axes, "T"))
+
+                    # Determine the metadata values that require reading the image data
+                    try:
+                        im_arr = page.asarray()
+                        metadata["num_unique_values"].append(len(np.unique(im_arr)))
```

Comment:
@kostrykin has this turned out to be useful ? `page.asarray()` may read in multi gigabyte files, causing the set metadata process to be killed or run out of memory. Do you have an idea how this can be done in a less memory intensive way ?

Reply:
FTR https://github.com/galaxyproject/galaxy/pull/19760#issuecomment-2706084196